### PR TITLE
Add new storage interfaces

### PR DESCRIPTION
**Pull request by stgraber**
This puts some of the infrastructure in place for the replacement storage API.

### CODE REVIEW

**Comment by stgraber** [comment]
@tomponline @monstermunchkin I tried to get all the interfaces/structs in order but this isn't currently doable, due to at least:
 - `backup` being referenced, should either be moved to its own package, or become an interface or maybe both
 - All the migration structs being in the `main` package due to dependencies on backups and storage. Once backup is sorted, we may be able to untangle the 3 bits of `migration` that we need and move them over to the `migration` package. For now, I've put empty stubs in there so that we can focus on the rest of the logic.

**Comment by stgraber** [comment]
One interesting challenge with this will be to keep that `drivers` package separate from the `storage` package. So far it looks quite possible as they're really each acting at a different level, but we'll see if I missed something :)

I ended up calling the main storage backend `lxdBackend`, that's mostly because we had to pick a name and this opens up the possibility of maybe supporting storage backends from other projects at some point in the distant future :)

**Comment by stgraber** [comment]
The `Driver` interface in there is just an early draft, we may well need to update a bunch of those functions and it's missing quite a bunch more functions we'll need to add.

**Comment by stgraber** [comment]
@monstermunchkin per above, it'd be good if we could sort out the `migration` package, it's currently mostly held by the old `storage` interface which in turn cannot be partially duplicated inside `migration` due to it depending on the `backup` struct that's in main.

Can you take a look at `backup` on Monday, see if it's something that'd make sense to move to its own package at this stage or if we can re-shuffle it to be an interface instead so we can untangle things a bit?

I have most of the patch needed to move a bunch of stuff into `migration` so we unblock `SinkArgs`, `SourceArgs` and `StorageSourceDriver` which I can resuscitate and push once `backup` is unblocked.

**Comment by stgraber** [comment]
@tomponline I've implemented a few bits of the `dir` driver here and there as a starting point.

A few things I ended up having to sort out:
 - Calling `CreatePool` now takes a pointer to api.StoragePool, allowing the backing driver to modify the Config. That's used to set the `source` property when not specified (for dir) but will also be used to similarly set a default size and loop device file for those backends that support that.
 - It is assumed that a successful call to `CreatePool` will be followed by a call to `Mount()`, similarly all pools should get loaded with `GetPoolByName` and then `Mount()` called on them during daemon startup. Being able to assume this is done properly will save us from having to do a LOT of internal checks for whether we're mounted or not.
 - The general idea is that we shouldn't have any driver-specific knowledge in the `storage` package, this should all be self-contained in the `drivers` package. This includes validating configuration (we'll want a list of valid keys and validators for each driver).
 - I've been trying to keep `dir` very simple, moving bits we expect to be shared with at least one other storage driver into `utils`.

File: lxd/storage/drivers/interface.go [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+)
+
+// VolumeType represents a storage volume type.
+type VolumeType string
+
+// VolumeTypeCustom represents a custom storage volume.
+const VolumeTypeCustom = VolumeType("custom")
+
+// VolumeTypeContainer represents a container storage volume.
+const VolumeTypeContainer = VolumeType("containers")
```
Comment (tomponline) [inline lxd/storage/drivers/interface.go]: @stgraber I wondered why these constants are singular but the string value of them is plural? Is this for historical reasons?

**Comment by tomponline** [comment]
@stgraber thanks for this, I've been reading through the code and your comments this morning.

One question I do have is regarding the `lxdBackend` struct and its purpose. You mention that the intention was to keep the storage package separate from the drivers package, such that we don't expose driving specific logic into the wider LXD. This makes sense and is achieved by having `CreatePool` return a `Pool` interface.

However I am confused why the `lxdBackend` cannot be used as an embedded struct in each specific driver implementation, called `driverCommon` or similar, like `deviceCommon` is in the `device` package, this would provide common implementations for creating directories etc that are shared between all storage drivers. I can see that the current implementation of `lxdBackend` seems to be responsible for just creation and deletion of the the directories needed for a storage pool on an LXD host.

Your earlier comments mention the future need of supporting non-LXD backends, so perhaps this is why there is a needed for the `lxdBackend` struct and a pointer to the driver being stored within it. 

Could you help me understand why the `lxdBackend` is needed? Thanks

**Pull request closed without merging**